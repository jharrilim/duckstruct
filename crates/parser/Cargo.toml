[package]
name = "parser"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lexer = { path = "../lexer" }
syntax = { path = "../syntax" }
drop_bomb = "0.1.5"
project-root = "0.2.2"
rowan = "0.15.4"
num-derive = "0.3.3"
num-traits = "0.2.14"
text-size = "1.0.0"

[dev-dependencies]
insta = { version = "1.21.0", features = ["yaml", "serde"] }
serde = { version = "1.0.145", features = ["derive"] }
