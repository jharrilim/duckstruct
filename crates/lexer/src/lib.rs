pub mod token;

use std::ops::Range;

use logos::Logos;
use text_size::{TextRange, TextSize};
use token::TokenKind;

pub struct Lexer<'a> {
  inner: logos::Lexer<'a, TokenKind>,
}

impl<'a> Lexer<'a> {
  pub fn new(input: &'a str) -> Self {
    Self {
      inner: TokenKind::lexer(input),
    }
  }
}

impl<'a> Iterator for Lexer<'a> {
  type Item = Token<'a>;

  fn next(&mut self) -> Option<Self::Item> {
    let kind = self.inner.next()?;
    let text = self.inner.slice();
    let line = self.inner.extras.line_heads.len();
    let line_index = *self.inner.extras.line_heads.last().unwrap_or(&0);

    let range = {
      let Range { start, end } = self.inner.span();
      let start = start - line_index;
      let end = end - line_index;

      let start = TextSize::try_from(start).unwrap();
      let end = TextSize::try_from(end).unwrap();

      TextRange::new(start, end)
    };

    Some(Self::Item {
      kind,
      text,
      range,
      line,
    })
  }
}

#[derive(Debug, PartialEq, Eq)]
pub struct Token<'a> {
  pub kind: TokenKind,
  pub text: &'a str,
  pub range: TextRange,
  pub line: usize,
}
