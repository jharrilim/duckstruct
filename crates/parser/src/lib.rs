use self::parser::{Parse, Parser};
use lexer::Lexer;
use sink::Sink;

mod event;
mod expressions;
mod marker;
mod operators;
mod parse_error;
pub mod parser;
mod parsers;
mod sink;
mod source;
mod statements;

pub fn parse(input: &str) -> Parse {
  let tokens: Vec<_> = Lexer::new(input).collect();
  let parser = Parser::new(&tokens);
  let events = parser.parse();
  let sink = Sink::new(&tokens, events);

  sink.finish()
}

#[cfg(test)]
mod tests {
  use super::*;
  use serde::Serialize;

  #[derive(Serialize)]
  struct ParseInfo {
    code: &'static str,
  }

  macro_rules! parse_snapshot {
    ($code:expr) => {
        let parse_info = ParseInfo { code: $code };
        let parse = parse($code);
        insta::with_settings!({
          info => &parse_info,
        }, {
          insta::assert_snapshot!(&parse.debug_tree());
        });
    };
  }

  #[test]
  fn parse_nothing() {
    let parse = parse("");
    assert_eq!(parse.debug_tree(), r#"Root@0..0"#);
  }

  #[test]
  fn parse_number() {
    let code = "123";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_variable_ref() {
    let code = "aVariableName";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_simple_infix_expression() {
    let code = "1+2";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_negation() {
    let code = "-10";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_nested_parentheses() {
    let code = "((((10))))";
    parse_snapshot!(code);
  }

  #[test]
  fn parentheses_affect_precedence() {
    let code = "5*(2+1)";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_number_preceded_by_whitespace() {
    let code = "   9876";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_number_followed_by_whitespace() {
    let code = "999   ";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_number_surrounded_by_whitespace() {
    let code = " 123     ";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_comment() {
    let code = "// hello!";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_boolean_literal_true() {
    let code = "true";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_boolean_literal_false() {
    let code = "false";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_string_literal() {
    let code = "\"hello\"";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_string_literal_empty() {
    let code = "\"\"";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_let_statement() {
    let code = "let x = 100;";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_destructuring() {
    let code = "let { x, y } = x;";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_destructuring_field_alias() {
    let code = "let { x: asd, y } = z;";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_array_destructuring() {
    // might want to treat _ as a special ident at some point
    let code = "let [_, y] = x;";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_nested_patterns() {
    let code = "let [{ x, }, z] = weel;";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_function_expression() {
    let code = r#"
      f double(x) = x * 2;
    "#;
    parse_snapshot!(code);
  }

  #[test]
  fn parse_anonymous_function_expression() {
    let code = r#"
      f(x) = x / 2;
    "#;
    parse_snapshot!(code);
  }

  #[test]
  fn parse_function() {
    let code = r#"
      f upper(x) {
        let y = x * 5;
        y + 100
      }
    "#;
    parse_snapshot!(code);
  }

  #[test]
  fn parse_function_definition_with_multiple_arguments() {
    let code = r#"
      f add(x, y) = x + y;
    "#;
    parse_snapshot!(code);
  }

  #[test]
  fn parse_let_expression_followed_by_f_definition() {
    let code = r#"
      let x = 10;
      f y(x) = x * 2;
    "#;
    parse_snapshot!(code);
  }

  #[test]
  fn parse_if_condition() {
    let code = r#"
      if x > 0 {
        x + 1
      }
    "#;
    parse_snapshot!(code);
  }

  #[test]
  fn parse_if_condition_with_else() {
    let code = r#"
      if x > 0 {
        x + 1
      } else {
        x - 1
      }
    "#;
    parse_snapshot!(code);
  }

  #[test]
  fn parse_and_expression() {
    let code = "x && y";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_or_expression() {
    let code = "x || y";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_conditional_assignment() {
    let code = "let x = if true { 1 } else { 2 }";
    parse_snapshot!(code);
  }

  #[test]
  fn err_let_stmt() {
    let code = "let x =\nlet y = 1";
    let parsed = parse(code);
    assert_eq!(parsed.errors.len(), 1);
  }

  #[test]
  fn parse_function_call() {
    let code = "foo()";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_function_call_with_argument() {
    let code = "foo(1)";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_immediately_invoked_function() {
    let code = "(f() = { 1 })()";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_curried_function_let_assignment() {
    let code = "let add = f(x) = f(y) = x + y";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_curried_function_invocation() {
    let code = "add(1)(2)";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_empty_array_expression() {
    let code = "[]";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_array_expression() {
    let code = "[1, 2, 3]";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_two_dimensional_array_expression() {
    let code = "[[1, 2], [3, 4]]";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_array_expression_with_trailing_comma() {
    let code = "[1, 2, 3,]";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_funky_array_expression() {
    let code = "[(1 + 4), \"h[e]llo\", { let x = 1; x + 1 }]";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_array_assignment() {
    let code = "let x = [1, 2, 3];";
    parse_snapshot!(code);
  }

  #[test]
  fn parse_array_assignment_with_destructuring() {
    let code = "let [x, y] = [1, 2];";
    parse_snapshot!(code);
  }
}
