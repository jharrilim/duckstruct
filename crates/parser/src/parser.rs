use lexer::{token::TokenKind, Token};

use syntax::{SyntaxKind, SyntaxNode};

use rowan::GreenNode;

use crate::statements::stmt;

use super::{event::Event, marker::Marker, source::Source};

pub struct Parser<'l, 'input> {
  pub(crate) source: Source<'l, 'input>,
  pub(crate) events: Vec<Event>,
  #[allow(unused)]
  pub(crate) errors: Vec<String>,
}

impl<'l, 'input> Parser<'l, 'input> {
  pub(crate) fn new(tokens: &'l [Token<'input>]) -> Self {
    Self {
      source: Source::new(tokens),
      events: Vec::new(),
      errors: Vec::new(),
    }
  }

  pub(crate) fn parse(mut self) -> Vec<Event> {
    let root_marker = self.start();
    while self.source.peek_kind().is_some() {
      stmt(&mut self);
    }
    root_marker.complete(&mut self, SyntaxKind::Root);

    self.events
  }

  pub(crate) fn start(&mut self) -> Marker {
    let pos = self.events.len();
    self.events.push(Event::Placeholder);

    Marker::new(pos)
  }

  pub(crate) fn peek(&mut self) -> Option<TokenKind> {
    self.source.peek_kind()
  }

  // Consume the current token and add it to the AST
  pub(crate) fn bump(&mut self) {
    self.source.next_token().unwrap();
    self.events.push(Event::AddToken);
  }

  pub(crate) fn at(&mut self, kind: TokenKind) -> bool {
    self.peek() == Some(kind)
  }
}

pub struct Parse {
  pub root: GreenNode,
}

impl Parse {
  pub fn debug_tree(&self) -> String {
    let syntax_node = SyntaxNode::new_root(self.root.clone());
    let formatted = format!("{:#?}", syntax_node);

    // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
    formatted[0..formatted.len() - 1].to_string()
  }

  pub fn syntax(&self) -> SyntaxNode {
    SyntaxNode::new_root(self.root.clone())
  }
}