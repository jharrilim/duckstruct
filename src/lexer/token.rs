use logos::Logos;
use num_derive::{FromPrimitive, ToPrimitive};

#[derive(Logos, Debug, PartialEq, Clone, Copy, FromPrimitive, ToPrimitive)]
pub enum SyntaxKind {
    #[token("f")]
    Function,

    #[token("let")]
    Let,

    #[token("class")]
    Class,

    #[token("{")]
    LeftBrace,

    #[token("}")]
    RightBrace,

    #[token("(")]
    LeftParenthesis,

    #[token(")")]
    RightParenthesis,

    #[token(".")]
    Period,

    #[regex(r"[0-9]+")]
    Number,

    #[token("+")]
    Plus,

    #[token("-")]
    Minus,

    #[token("*")]
    Asterisk,

    #[token("/")]
    ForwardSlash,

    #[token("=")]
    Equals,

    #[regex(r#""([^\\"]|\.)*""#)]
    String,

    #[token("\"")]
    DoubleQuote,

    #[token("'")]
    SingleQuote,

    #[token(";")]
    Semicolon,

    #[token(":")]
    Colon,

    #[regex(r"[a-zA-Z_][a-zA-Z\d_]+")]
    Identifier,

    #[regex(r"[\s]+", logos::skip)]
    Whitespace,

    #[error]
    Error,

    Root,
    BinaryOperation,
}

#[cfg(test)]
mod tests {
    use crate::lexer::Lexer;

    use super::*;
    use project_root::get_project_root;
    use std::fs::read_to_string;
    use std::path::Path;

    #[test]
    fn lexes_function_definition() {
        let code = r#"
            f bark() {

            }
        "#;
        let mut lex = Lexer::new(code);

        assert_eq!(lex.next(), Some((SyntaxKind::Function, "f")));
        assert_eq!(lex.next(), Some((SyntaxKind::Identifier, "bark")));
        assert_eq!(lex.next(), Some((SyntaxKind::LeftParenthesis, "(")));
        assert_eq!(lex.next(), Some((SyntaxKind::RightParenthesis, ")")));
        assert_eq!(lex.next(), Some((SyntaxKind::LeftBrace, "{")));
        assert_eq!(lex.next(), Some((SyntaxKind::RightBrace, "}")));
    }

    #[test]
    fn lexes_goal_input() {
        let root = get_project_root().unwrap();
        let input_path = Path::new("goal/input.ds");
        let f = read_to_string(root.join(input_path).as_path()).unwrap();
        let mut lex = SyntaxKind::lexer(&f);

        assert_eq!(lex.next(), Some(SyntaxKind::Class));
        assert_eq!(lex.next(), Some(SyntaxKind::Identifier));
        assert_eq!(lex.next(), Some(SyntaxKind::LeftBrace));
        assert_eq!(lex.next(), Some(SyntaxKind::Identifier));
        assert_eq!(lex.next(), Some(SyntaxKind::Equals));
        assert_eq!(lex.next(), Some(SyntaxKind::String));

        assert_eq!(lex.next(), Some(SyntaxKind::Semicolon));
    }
}
