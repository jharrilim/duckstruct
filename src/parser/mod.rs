mod event;
mod expressions;
mod marker;
mod operators;
mod parse_error;
mod parsers;
mod sink;
mod source;

use lexer::{token::TokenKind, Lexer, Token};

use syntax::{SyntaxKind, SyntaxNode};

use rowan::GreenNode;
use sink::Sink;

use self::{event::Event, expressions::expr, marker::Marker, source::Source};

struct Parser<'l, 'input> {
  source: Source<'l, 'input>,
  events: Vec<Event>,
}

impl<'l, 'input> Parser<'l, 'input> {
  fn new(tokens: &'l [Token<'input>]) -> Self {
    Self {
      source: Source::new(tokens),
      events: Vec::new(),
    }
  }

  fn parse(mut self) -> Vec<Event> {
    let root_marker = self.start();
    expr(&mut self);
    root_marker.complete(&mut self, SyntaxKind::Root);

    self.events
  }

  fn start(&mut self) -> Marker {
    let pos = self.events.len();
    self.events.push(Event::Placeholder);

    Marker::new(pos)
  }

  fn peek(&mut self) -> Option<TokenKind> {
    self.source.peek_kind()
  }

  // Consume the current token and add it to the AST
  fn bump(&mut self) {
    self.source.next_token().unwrap();
    self.events.push(Event::AddToken);
  }

  fn at(&mut self, kind: TokenKind) -> bool {
    self.peek() == Some(kind)
  }
}

pub(crate) fn parse(input: &str) -> Parse {
  let tokens: Vec<_> = Lexer::new(input).collect();
  let parser = Parser::new(&tokens);
  let events = parser.parse();
  let sink = Sink::new(&tokens, events);

  Parse {
    green_node: sink.finish(),
  }
}

pub(crate) struct Parse {
  green_node: GreenNode,
}

impl Parse {
  pub fn debug_tree(&self) -> String {
    let syntax_node = SyntaxNode::new_root(self.green_node.clone());
    let formatted = format!("{:#?}", syntax_node);

    // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
    formatted[0..formatted.len() - 1].to_string()
  }
}


#[cfg(test)]
mod tests {
  use super::*;
  use expect_test::{expect, Expect};

  fn check(input: &str, expected_tree: Expect) {
    let parse = parse(input);
    expected_tree.assert_eq(&parse.debug_tree());
  }

  #[test]
  fn parse_nothing() {
    check("", expect![[r#"Root@0..0"#]])
  }

  #[test]
  fn parse_number() {
    check(
      "123",
      expect![[r#"
          Root@0..3
            Literal@0..3
              Number@0..3 "123""#]],
    );
  }

  #[test]
  fn parse_variable_ref() {
    check(
      "aVariableName",
      expect![[r#"
          Root@0..13
            VariableReference@0..13
              Identifier@0..13 "aVariableName""#]],
    );
  }

  #[test]
  fn parse_simple_infix_expression() {
    check(
      "1+2",
      expect![[r#"
          Root@0..3
            InfixExpression@0..3
              Literal@0..1
                Number@0..1 "1"
              Plus@1..2 "+"
              Literal@2..3
                Number@2..3 "2""#]],
    );
  }

  #[test]
  fn parse_negation() {
    check(
      "-10",
      expect![[r#"
          Root@0..3
            PrefixExpression@0..3
              Minus@0..1 "-"
              Literal@1..3
                Number@1..3 "10""#]],
    );
  }

  #[test]
  fn parse_nested_parentheses() {
    check(
      "((((10))))",
      expect![[r#"
          Root@0..10
            ParenExpression@0..10
              LeftParenthesis@0..1 "("
              ParenExpression@1..9
                LeftParenthesis@1..2 "("
                ParenExpression@2..8
                  LeftParenthesis@2..3 "("
                  ParenExpression@3..7
                    LeftParenthesis@3..4 "("
                    Literal@4..6
                      Number@4..6 "10"
                    RightParenthesis@6..7 ")"
                  RightParenthesis@7..8 ")"
                RightParenthesis@8..9 ")"
              RightParenthesis@9..10 ")""#]],
    );
  }

  #[test]
  fn parentheses_affect_precedence() {
    check(
      "5*(2+1)",
      expect![[r#"
          Root@0..7
            InfixExpression@0..7
              Literal@0..1
                Number@0..1 "5"
              Asterisk@1..2 "*"
              ParenExpression@2..7
                LeftParenthesis@2..3 "("
                InfixExpression@3..6
                  Literal@3..4
                    Number@3..4 "2"
                  Plus@4..5 "+"
                  Literal@5..6
                    Number@5..6 "1"
                RightParenthesis@6..7 ")""#]],
    );
  }

  #[test]
  fn parse_number_preceded_by_whitespace() {
    check(
      "   9876",
      expect![[r#"
          Root@0..7
            Whitespace@0..3 "   "
            Literal@3..7
              Number@3..7 "9876""#]],
    );
  }

  #[test]
  fn parse_number_followed_by_whitespace() {
    check(
      "999   ",
      expect![[r#"
          Root@0..6
            Literal@0..6
              Number@0..3 "999"
              Whitespace@3..6 "   ""#]],
    );
  }

  #[test]
  fn parse_number_surrounded_by_whitespace() {
    check(
      " 123     ",
      expect![[r#"
          Root@0..9
            Whitespace@0..1 " "
            Literal@1..9
              Number@1..4 "123"
              Whitespace@4..9 "     ""#]],
    );
  }

  #[test]
  fn parse_comment() {
    check(
      "// hello!",
      expect![[r#"
              Root@0..9
                Comment@0..9 "// hello!""#]],
    );
  }

  #[test]
  fn parse_let_statement() {
    check(
      "let x = 100;",
      expect![[r#"
          Root@0..11
            LetExpression@0..11
              Let@0..3 "let"
              Whitespace@3..4 " "
              Assignment@4..6
                Identifier@4..5 "x"
                Whitespace@5..6 " "
              Equals@6..7 "="
              Whitespace@7..8 " "
              Literal@8..11
                Number@8..11 "100""#]]
    )
  }

  #[test]
  fn parse_destructuring() {
    check(
      "let { x, y } = x;",
      expect![[r#"
          Root@0..16
            LetExpression@0..16
              Let@0..3 "let"
              Whitespace@3..4 " "
              Assignment@4..13
                StructPattern@4..13
                  LeftBrace@4..5 "{"
                  Whitespace@5..6 " "
                  Identifier@6..7 "x"
                  Comma@7..8 ","
                  Whitespace@8..9 " "
                  Identifier@9..10 "y"
                  Whitespace@10..11 " "
                  RightBrace@11..12 "}"
                  Whitespace@12..13 " "
              Equals@13..14 "="
              Whitespace@14..15 " "
              VariableReference@15..16
                Identifier@15..16 "x""#]]
    )
  }

  #[test]
  fn parse_destructuring_field_alias() {
    check(
      "let { x: asd, y } = z;",
      expect![[r#"
          Root@0..21
            LetExpression@0..21
              Let@0..3 "let"
              Whitespace@3..4 " "
              Assignment@4..18
                StructPattern@4..18
                  LeftBrace@4..5 "{"
                  Whitespace@5..6 " "
                  Identifier@6..7 "x"
                  Colon@7..8 ":"
                  Whitespace@8..9 " "
                  Identifier@9..12 "asd"
                  Comma@12..13 ","
                  Whitespace@13..14 " "
                  Identifier@14..15 "y"
                  Whitespace@15..16 " "
                  RightBrace@16..17 "}"
                  Whitespace@17..18 " "
              Equals@18..19 "="
              Whitespace@19..20 " "
              VariableReference@20..21
                Identifier@20..21 "z""#]]
    )
  }
}
