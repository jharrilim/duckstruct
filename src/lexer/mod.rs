pub mod token;

use logos::Logos;
use token::SyntaxKind;

pub(crate) struct Lexer<'a> {
    inner: logos::Lexer<'a, SyntaxKind>,
}

impl<'a> Lexer<'a> {
    pub(crate) fn new(input: &'a str) -> Self {
        Self {
            inner: SyntaxKind::lexer(input),
        }
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = (SyntaxKind, &'a str);

    fn next(&mut self) -> Option<Self::Item> {
        let kind = self.inner.next()?;
        let text = self.inner.slice();

        Some((kind, text))
    }
}

