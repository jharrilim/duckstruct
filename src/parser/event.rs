use crate::lexer::token::SyntaxKind;

#[derive(Debug, Clone, PartialEq)]
pub(super) enum Event {
  StartNode {
    kind: SyntaxKind,
    forward_parent: Option<usize>,
  },
  AddToken {
    kind: SyntaxKind,
    text: String,
  },
  FinishNode,
  Placeholder,
}
