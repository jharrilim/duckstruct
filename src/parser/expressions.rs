use super::{marker::CompletedMarker, operators::InfixOp, parsers};
use crate::parser::Parser;
use lexer::token::TokenKind;
use syntax::SyntaxKind;

pub(super) fn expr(p: &mut Parser) {
  expr_binding_power(p, 0)
}

pub(super) fn expr_binding_power(p: &mut Parser, minimum_binding_power: u8) {
  let mut lhs = if let Some(lhs) = lhs(p) {
    lhs
  } else {
    return;
  };

  loop {
    let op = match p.peek() {
      Some(TokenKind::Plus) => InfixOp::Add,
      Some(TokenKind::Minus) => InfixOp::Sub,
      Some(TokenKind::Asterisk) => InfixOp::Mul,
      Some(TokenKind::ForwardSlash) => InfixOp::Div,
      _ => return,
    };

    let (left_binding_power, right_binding_power) = op.binding_power();

    if left_binding_power < minimum_binding_power {
      return;
    }

    p.bump();

    let m = lhs.precede(p);
    expr_binding_power(p, right_binding_power);
    lhs = m.complete(p, SyntaxKind::InfixExpression);
  }
}

fn lhs(p: &mut Parser) -> Option<CompletedMarker> {
  let completed_marker = match p.peek() {
    Some(TokenKind::Number) => parsers::literal(p),
    Some(TokenKind::Identifier) => parsers::variable_ref(p),
    Some(TokenKind::Minus) => parsers::prefix_expr(p),
    Some(TokenKind::LeftParenthesis) => parsers::paren_expr(p),
    _ => return None,
  };
  Some(completed_marker)
}
