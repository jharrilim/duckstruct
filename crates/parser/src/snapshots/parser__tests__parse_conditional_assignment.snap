---
source: crates/parser/src/lib.rs
expression: "&parse.debug_tree()"
info:
  code: "let x = if true { 1 } else { 2 }"
---
Root@0..32
  LetStatement@0..32
    Let@0..3 "let"
    Whitespace@3..4 " "
    Identifier@4..5 "x"
    Whitespace@5..6 " "
    Equals@6..7 "="
    Whitespace@7..8 " "
    ConditionalExpression@8..32
      If@8..10 "if"
      Whitespace@10..11 " "
      ConditionalPredicate@11..16
        Boolean@11..16
          Boolean@11..15 "true"
          Whitespace@15..16 " "
      LeftBrace@16..17 "{"
      Whitespace@17..18 " "
      IfCondition@18..20
        Number@18..20
          Number@18..19 "1"
          Whitespace@19..20 " "
      RightBrace@20..21 "}"
      Whitespace@21..22 " "
      Else@22..26 "else"
      Whitespace@26..27 " "
      LeftBrace@27..28 "{"
      Whitespace@28..29 " "
      ElseCondition@29..31
        Number@29..31
          Number@29..30 "2"
          Whitespace@30..31 " "
      RightBrace@31..32 "}"
